zilliqa/src/contracts/deposit.sol

```diff
@@ -138,61 +207,238 @@ contract Deposit {
+    }
+
+    function tickEpoch() public {
+        // if current epoch has already been processed, exit
+        if (headProcessedEpoch >= currentEpoch()) {
+            return;
+        }
+
+        uint8 queueIndex = currentQueueIndex();
+
+        // Process pending queue for this epoch
+        for (uint i = 0; i < pendingQueues[queueIndex].length(); ++i) {
+            Value storage newDeposit = pendingQueues[queueIndex].popFront();
+            _stakersMap[newDeposit.key].stakedBalance += newDeposit.amount;
+            totalStake += newDeposit.amount;
+            // If the validator wasn't already active (or suspended or slashed), activate it
+            if (_stakersMap[newDeposit.key].status == Status.Pending) {
+                _stakersMap[newDeposit.key].status = Status.Active;
+                _stakersMap[newDeposit.key].keyIndex = int(committeeKeys.length);
+                committeeKeys.push(newDeposit.key);
+            }
+        }
+
+        // Process unstakings queue for this epoch
+        for (uint i = 0; i < unstakingQueues[queueIndex].length(); ++i) {
+            Value storage unstaking = unstakingQueues[queueIndex].popFront();
+            _stakersMap[unstaking.key].stakedBalance -= unstaking.amount;
+            _stakersMap[unstaking.key].unstakingBalance -= unstaking.amount;
+            _stakersMap[unstaking.key].freeBalance += unstaking.amount;
+            totalStake -= unstaking.amount;
+
+            // If the validator's stakedBalance is now below minimum stake, suspend it
+            if (_stakersMap[unstaking.key].stakedBalance < minimumStake && _stakersMap[unstaking.key].status == Status.Active) {
+                _stakersMap[unstaking.key].status = Status.Suspended;
+                swapRemoveValidator(uint(_stakersMap[unstaking.key].keyIndex)); // should be safe, because it should not be possible for a validator to be Active without having a positive keyIndex
+            }
+        }
+
+        // Process kick queue - suspensions (voluntary, and for liveness violations) and slashings (for consensus violations)
+        for (uint i = 0; i < kickQueue.length(); ++i) {
+            Value storage kick = kickQueue.popFront();
+            Staker storage staker = _stakersMap[kick.key];
+            // Remove the staker from the total stake and the active committee
+            totalStake -= staker.stakedBalance;
+            // Only remove the staker from the committee if we know it's in it currently
+            // There are edge cases where it may not be necessary - e.g. if a withdrawal is processed on the same epoch as a slash/suspend, it will already have been removed just above
+            if (staker.keyIndex >= 0) {
+                swapRemoveValidator(uint(staker.keyIndex));
+            }
+
+            // Apply penalty, if any (might be 0 e.g. for voluntary suspension)
+            // TODO: collect these rewards somewhere? Right now they're effectively burned
```

Comment:
a fraction of the slashing penalty shall be paid as whistleblower reward to whoever submitted the slashing evidence, the rest must be burned back to the zero account